File: 2025/day11-2.hs
import Data.Bifunctor qualified
import Data.Map qualified
import Data.Maybe (fromJust)

type Devices = Data.Map.Map String [String]

type Cache = Data.Map.Map (String, Bool, Bool) Int

solution :: String -> Int
solution content = fst (count Data.Map.empty (Data.Map.fromList . map parse . lines $ content) "svr" False False)

count :: Cache -> Devices -> String -> Bool -> Bool -> (Int, Cache)
count cache _ "out" True True = (1, cache)
count cache _ "out" _ False = (0, cache)
count cache _ "out" False _ = (0, cache)
count cache ds name fft dac
  | Data.Map.member key cache = (fromJust . Data.Map.lookup key $ cache, cache)
  | otherwise = Data.Bifunctor.second (Data.Map.insert key (fst result)) result
  where
    key = (name, fft, dac)
    result =
      foldl
        (\acc target -> helper acc (count (snd acc) ds target (fft || name == "fft") (dac || name == "dac")))
        (0, cache)
        . fromJust
        . Data.Map.lookup name
        $ ds

helper :: (Int, Cache) -> (Int, Cache) -> (Int, Cache)
helper (a1, c1) (a2, c2) = (a1 + a2, Data.Map.union c1 c2)

parse :: String -> (String, [String])
parse line =
  ( init . head . words $ line,
    tail . words $ line
  )

main :: IO ()
main = do readFile "day11.dat" >>= print . solution

File: 2025/day11_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Devices = dict[str, frozenset[str]]


def solution(content: list[str]) -> Any:
    return count(
        {key: value for key, value in (parse(line) for line in content)},
        "svr",
        False,
        False,
    )


cache: dict[tuple[str, bool, bool], int] = {}


def count(deivces: Devices, name: str, fft: bool, dac: bool) -> int:
    if name == "out":
        return 1 if fft and dac else 0
    key = (name, fft, dac)
    if key in cache:
        return cache[key]
    result = sum(
        count(deivces, target, fft or (name == "fft"), dac or (name == "dac"))
        for target in deivces[name]
    )
    cache[key] = result
    return result


def parse(line: str) -> tuple[str, frozenset[str]]:
    first, *rest = line.split(" ")
    return (first[:-1], frozenset(item.strip() for item in rest))


def main() -> None:
    with open("day11.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
